services/utils.py

```diff
@@ -26,3 +29,33 @@ def get_service_from_service_key(request, raise_exception: bool = True) -> Servi
             raise NotAuthenticated()
 
     return service
+
+
+def get_service_from_request(
+    request, raise_exception: bool = True
+) -> Optional[Service]:
```

Comment:
Hmm I'm a bit confused about this, because it both sets the service for the request and at the same time it returns the instance that has been connected to the request.

So, for example, you could do
```python
request.service = get_service_from_request(request)
```
and it would set the value twice.

IMO, to keep the matters clean, we should either:
* rename this one to `set_service_for_request`
* skip the setting of `request.client_id/request.service` and just return the instance of `Service/ServiceClientId`, then do the setting in a more explicit place

Reply:
Setting the `service` to the request here currently only serves the purpose of caching the service so that the logic for fetching the service only needs to run once per request. This reduces the queries the need to hit the DB. The function is currently called a few times per request. When using `ApiTokenAuthentication` we can't afaik have this logic in Django middleware since DRF authentication modules are run later.

Reply:
Ok I think now it makes a bit more sense, using the request to cache the service instance sounds good 👍 I guess we don't really need the client ID other than for identifying the service, so +1 to getting rid of it.